sections/advanced/reaching-ui-state.md

```diff
@@ -0,0 +1,222 @@
+### One Paragraph Explainer
+
+It adds value to cover a UI scenario *once*, and it provides little value to duplicate any parts of it in other tests; tests which may require a relevant state of the system. Suppose in a new test you require a state, and that state -partially or in full- duplicates testing from a UI test. Scenarios like this are excellent for utilizing certain techniques:
+
+* Direct navigation
+* Network stub record & play
+* Application actions
+* Seeding the database
+
+> Disclaimer: applications of the whole package of these techniques are only possible in Cypress (as far as we know), consequently the code samples below are are in Cypress context.
+
+<br/><br/>
+
+### Direct navigation
+
+This is the easiest technique and is applicable in any framework. Suppose the intent of the test is concerned with a certain page in your application. Instead of click navigating, directly visit the URL. Upon landing you can either wait for a UI element (any test framework) or network calls (some test frameworks), or both.
+
+```javascript
+// Test A covers click-navigation to a certain page.
+// This is Test B, and navigating to that page is the prerequisite step.
+
+// assuming baseUrl is set in cypress.json or config file
+// directly navigate to the page.
+cy.visit('/endpoint');
+
+// to ensure stability, wait for network (preferred), ui elements, or both
+
+// note: you can optionally add a url check
+cy.url().should('contain', 'endpoint');
+// cy.url().should('match', /endpoint/); // there are many, some more complex, ways of doing it
+
+
+// network wait:
+cy.server();  // required to use cy.route
+// usually a GET request. Is aliased so we can wait for it.
+cy.route('some-xhr-call-that-happens-upon-landing').as('crutcXHR');
+cy.wait('@crutchXHR', {timeout: 15000}); // can add an optional timeout if the page takes too long
+
+// ui-element wait is straightforward, and may be optional, as well as less stable)
+cy.get('element-on-page').should('exist').and('be.visible');
+
+```
+
+<br/><br/>
+
+### Network stub record & play
+This is an advanced technique that strongly relates to UI-integration tests. Recall UI-integration references [1](../testing-strategy/component-vs-integration-vs-e2e-testing.md), [2](../real-life-examples/test-front-end-with-integration-back-end-with-e2e.md).
+
+Cypress allows to stub all network traffic. We can record the network data from an endpoint, and stub that response every time the UI makes a call to an arbitrary server.
+
+Start by copying the network data from devTools to a json file. Place it in `cypress/fixtures` folder. This folder is made for this purpose, and any reference to it will default to the root of the folder.
+
+![devtools > network tab](../../assets/images/ui-state/devtools-network.PNG)
+
+```javascript
+// json has been copied to a file `cypress/fixtures/agents.json`
+
+// this is a shorthand for cy.fixture(). More at https://docs.cypress.io/api/commands/fixture.html#Accessing-Fixture-Data
```

Comment:
This comment should be moved one line down, right?

Reply:
re-arranged the comment